ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java

```diff
@@ -100,11 +101,17 @@ public static boolean is_eligible_for_activation(BeaconState state, Validator va
    *     <a>https://github.com/ethereum/eth2.0-specs/blob/v0.8.0/specs/core/0_beacon-chain.md#get_active_validator_indices</a>
    */
   public static List<Integer> get_active_validator_indices(BeaconState state, UnsignedLong epoch) {
-    List<Validator> validators = state.getValidators();
-    return IntStream.range(0, validators.size())
-        .filter(index -> is_active_validator(validators.get(index), epoch))
-        .boxed()
-        .collect(Collectors.toList());
+    return BeaconStateWithCache.getTransitionCaches(state)
+        .getActiveValidators()
+        .get(
+            epoch,
+            e -> {
+              List<Validator> validators = state.getValidators();
+              return IntStream.range(0, validators.size())
+                  .filter(index -> is_active_validator(validators.get(index), epoch))
+                  .boxed()
+                  .collect(Collectors.toList());
+            });
```

Comment:
This works and I like how it winds up with very minimal changes, but I wonder if as we introduce more caches we'll wind up with a lot of boilerplate in getting the caches etc obscuring the actual business logic.

I think we should land this as-is, but I am wondering if we could be a bit radical here and take advantage of polymorphism.  https://github.com/PegaSysEng/artemis/commit/1144a564f684cd8d47c5d09a1323a83ab1a700e3 is what I have in mind.  If `BeaconState` provided a `getActiveValidatorIndices` method with the actual spec logic, then `BeaconStateWithCache` could just override that method to check the cache first.  Feels really nice and clean in many ways.

But there's a couple of downsides:
1. Not sure this scales well - we could wind up with half the transition code in `BeaconState`.
2. I don't see this playing well with hard forks after launch (neither does our current approach)

Which I think is enough to talk me out of it, and that it's better for classes like `ValidatorUtils` to use actual instances rather than all static methods - then this polymorphism approach could be used there, and different implementations could be provided for different hard forks, but that's a bigger scope of work.

You can safely ignore this for now if you like, I just keep pondering how we need to evolve the design to be more flexible in the future...  Happy to hear any thoughts anyone has though.

Reply:
Yep, I see what you mean and totally support this approach. The spec procedural style contradicts Java OOP. But we are still sticking to it to be closer to the original spec to easy follow spec changes. 
I think at some point this is absolutely must have, but suppose this should be a separate large refactoring and we could start it from your suggested change

Reply:
Yeah agreed, out of scope for this PR.